### PR TITLE
SCC: Minimal sync/hunt support to enable MacOS7 to boot on Macplus

### PR DESCRIPTION
**Pull request by JoakimLarsson**

### CODE REVIEW

File: src/devices/machine/z80scc.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ DONE (x) (p=partly)         NMOS         CMOS       ESCC      EMSCC
 // printf("TAG %lld %s%s Data:%d\n", machine().firstcpu->total_cycles(), __PRETTY_FUNCTION__, m_owner->tag(), data);
 
 #define VERBOSE 0
-#define LOGPRINT(x) do { if (VERBOSE) logerror x; } while (0)
+#define LOGPRINT(x) { do { if (VERBOSE) logerror x; } while (0); }
```
Comment (cuavas) [inline src/devices/machine/z80scc.cpp]: We prefer the form with `do { } while (0)` on the outside so the macro behaves like a function syntactically, requiring a following semicolon.

Comment (rb6502) [inline src/devices/machine/z80scc.cpp]: Maybe we should provide a logprint.h header with on/off and verbose-level forms so people don't keep reinventing that wheel?

**Comment by etabeta78** [comment]
Possibly with various verbosity levels, which is another thing which has
been re-implemented in dozen of drivers...

    f.

**Comment by rb6502** [comment]
eta: Yeah, that's what I meant by "verbose-level".  This stuff is *everywhere*.

**Comment by etabeta78** [comment]
whops I had misread your message :)

**Comment by mizapf** [comment]
I had better experience with topic-like trace flags instead of verbosity levels. You sometimes just want the log output of a particular feature, and not of everything in your driver. Also, assigning log levels in a suitable way is less easy that it sounds.
What I would conclude from the benchmarks is that a line like
if (SOME_LOG_FLAG) logerror(...) 
is not compiled at all when SOME_LOG_FLAG is set to 0, and thus does not have any impact on the code.

**Comment by rb6502** [comment]
@mizapf if you want to keep doing your own logging that goes beyond the simple cases we're discussing, you're allowed to.  But dozens of drivers would be covered by the proposal here and it would clean up and standardize a lot of usage.

File: src/devices/machine/z80scc.cpp (hunk 1 above)
Comment (JoakimLarsson) [inline src/devices/machine/z80scc.cpp]: After 30 years of programming it's not invention, rather lazynes from my side, I have a pattern I use when I develop stuff but I'd be more than happy to convert if we agree on something useful. These LOG printouts are part needed developing the device and part needed when using the device in a board driver as I try to hint about misuse etc. So once the device is complete many of the printout can be removed completely and once we get logerror to work onto something that can be copied and paste on all platforms I'd be glad to remove the redefinition to printf

The extra curly brackets is just a precaution in the case where a LOG is at an unbracketed  else clause as the original has no semicolon. I can of course add the semicolon instead.

I have 29 source files submitted so far and most of them uses this scheme, I suggest that I wait until we have an improved and agreed upon LOG print scheme before I convert and then that I do all of them, what do you think?

**Comment by JoakimLarsson** [comment]
I think it would be valuable to have some default memory map debug printouts that can be turned on in compile time that just prints read and writes in a standardized format with machine time stamps. very useful for documenting init sequences and see differences between roms

In addition a device could provide a set of names for registers and bits/bitmasks so that the output is human readable. This would speed up development of devices I think.

Another category of LOG printouts could be turned on and off in run time from the debugger, LUA or the code itself

And yes, I think a logprint.h would be great! :)

**Comment by mizapf** [comment]
I have no objections against log levels and of getting to some common way of usage; as I said, it's a matter of own experiences with logging. :)

**Comment by cuavas** [comment]
The point of the `do { ... } while (0)` is so that you *need* the semicolon, so it behaves syntactically like a function call.

This should work:
```
    whatever(...);
    LOG(...);
    whatever(...);
```

But this should not:
```
    whatever(...);
    LOG(...)
    whatever(...);
```

If you put in the extra braces, then it works without the semicolon and defeats the purpose of wrapping it in `do { ... } while (0)` altogether.

**Comment by cuavas** [comment]
The semicolon shouldn't be in the macro. It defeats the purpose of the do/while wrapper. This is standard practice across MAME. Please follow it.

**Comment by JoakimLarsson** [comment]
Ahh, ok I get your point. Let me change again., Thing is it made me look for a long time before I realized the problem.

**Pull request merged**